nixos/modules/security/acme.xml

```diff
@@ -66,4 +66,31 @@ options for the <literal>security.acme</literal> module.</para>
 
 </section>
 
+<section><title>Using ACME certificates in Nginx</title>
+<para>In practice ACME is mostly used for retrieval &amp; renewal of certificates that will be used in a webserver like Nginx. A configuration for Nginx that uses the certificates from ACME for <literal>foo.example.com</literal> will look similar to:</para>
+
+<programlisting>
+services.nginx.httpConfig = ''
+  server {
+    server_name foo.example.com;
+    listen 443 ssl;
+    ssl_certificate     ${config.security.acme.directory}/foo.example.com/fullchain.pem;
+    ssl_certificate_key ${config.security.acme.directory}/foo.example.com/key.pem;
+    root /var/www/foo.example.com/;
+  }
+'';
+</programlisting>
+
+<para>Now Nginx will try to use the certificates that will be retrieved by ACME. However, Nginx will not start until the certificates actually exist. Additionally, ACME will not be able to retrieve certificates until Nginx can host the challenge-files for ACME over HTTP. This results in a problem where the two services need eachother to function.</para>
+
+<para>To circumvent this problem we can generate self-signed certificates that Nginx can use until ACME has retrieved new ones. This can be done manually, but the ACME module can also do this automatically with the <literal>security.acme.preliminarySelfsigned</literal> option. If we use this option we need to make sure Nginx is only started after the self-signed certificate is generated. This can be configured using:</para>
+
+<programlisting>
+security.acme.preliminarySelfsigned = true;
+systemd.services.nginx = {
+  after = [ "acme-selfsigned-certificates.target" ];
```

Comment:
Yes, that's a possibility, but I can imagine that ACME could be for Httpd,
while Nginx is used for something else. In that case Nginx doesn't need to
be depending on `acme-selfsigned-certificates.target`.

That said, if the intention is to de-clutter the machine-configuration with
ACME+nginx specific workarounds, I agree. We might need to improve upon
this, but this PR already brings us a step further.

I'm still thinking of having explicit integration options.
`security.acme-integration.nginx.enable` or something similar. It could be
a separate module on top of ACME that makes use of the
`security.acme.preliminarySelfsigned` option. That way each webserver-acme
integration can implement their own additional options that are needed. For
nginx you would need to depend on `acme-selfsigned-certificates.target` and
watch for the certificate files to reload nginx. apache might already
reload when certificate files change, but for apache you could
automatically create the certificate definitions from
`services.httpd.virtualHosts.*.hostName`
and `services.httpd.virtualHosts.*.enableSSL` or maybe even automatic http
-> https redirects. It seems obvious to me to have such options when HTTPS
is becoming the default and ACME is becoming the default for HTTPS. At the
moment to use ACME correctly someone would need to know quite a bit about
how services are related to eachother.

On Mon, May 23, 2016 at 2:00 PM Domen Kožar <notifications@github.com>
wrote:

> In nixos/modules/security/acme.xml
> <https://github.com/NixOS/nixpkgs/pull/15562#discussion_r64210274>:
>
> > +    listen 443 ssl;
> > +    ssl_certificate     ${config.security.acme.directory}/foo.example.com/fullchain.pem;
> > +    ssl_certificate_key ${config.security.acme.directory}/foo.example.com/key.pem;
> > +    root /var/www/foo.example.com/;
> > +  }
> > +'';
> > +</programlisting>
> > +
> > +<para>Now Nginx will try to use the certificates that will be retrieved by ACME. However, Nginx will not start until the certificates actually exist. Additionally, ACME will not be able to retrieve certificates until Nginx can host the challenge-files for ACME over HTTP. This results in a problem where the two services need eachother to function.</para>
> > +
> > +<para>To circumvent this problem we can generate self-signed certificates that Nginx can use until ACME has retrieved new ones. This can be done manually, but the ACME module can also do this automatically with the <literal>security.acme.preliminarySelfsigned</literal> option. If we use this option we need to make sure Nginx is only started after the self-signed certificate is generated. This can be configured using:</para>
> > +
> > +<programlisting>
> > +security.acme.preliminarySelfsigned = true;
> > +systemd.services.nginx = {
> > +  after = [ "acme-selfsigned-certificates.target" ];
>
> Couldn't we put this into nginx itself? I think if target is missing in
> general, systemd will just skip it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/15562/files/350ee207b8c1ab2bf5ce2d4af709dc9398e5af3a#r64210274>
>